packages/kotti-ui/source/kotti-field-file-upload/KtFieldFileUploadRemote.vue

```diff
@@ -0,0 +1,246 @@
+		DropArea,
+		FileItemRemote,
+		KtField,
+		PreUploadedFileItem,
+		TakePhoto,
+	},
+	props: makeProps(KottiFieldFileUploadRemote.propsSchema),
+	setup(props, { emit }) {
+		const field = useField<KottiFieldFileUploadRemote.ValueInternal>({
+			emit,
+			/**
+			 * We consider this field empty if:
+			 * a. No files are selected, or
+			 * b. Value is invalid
+			 */
+			isEmpty: (value) => value.length === 0 || !isValidValue(value, 'Remote'),
+			props,
+			supports: KOTTI_FIELD_FILE_UPLOAD_SUPPORTS,
+		})
+
+		const newUploadedFiles = new Set<
+			KottiFieldFileUploadRemote.FileInfo['id']
+		>()
+		const preUploadedFiles = new Set<
+			KottiFieldFileUploadRemote.FileInfo['id']
+		>()
+
+		const filesList = computed<KottiFieldFileUploadRemote.FileInfo[]>(() =>
+			field.currentValue
+				.filter((fileItem) => !preUploadedFiles.has(fileItem.id))
+				.map((fileItem) => {
+					const fileInfo = buildFileInfo({
+						extensions: props.extensions,
+						fileItem,
+						isUploaded: newUploadedFiles.has(fileItem.id),
+						maxFileSize: props.maxFileSize,
+						progress: props.payload[fileItem.id]?.progress ?? 0,
+					})
+
+					if (fileItem.status !== fileInfo.status)
+						setStatus({ id: fileItem.id, status: fileInfo.status })
+
+					return fileInfo
+				}),
+		)
+		const showDropArea = computed(
+			() => props.allowMultiple || field.currentValue.length === 0,
+		)
+
+		const setStatus = (
+			payload: KottiFieldFileUploadRemote.Events.SetStatus,
+		) => {
+			field.setValue(
+				field.currentValue.map((fileItem) =>
+					fileItem.id === payload.id
+						? { ...fileItem, status: payload.status }
+						: fileItem,
+				),
+				{ forceUpdate: true },
+			)
+
+			if (payload.status === KottiFieldFileUploadRemote.Status.UPLOADED)
+				newUploadedFiles.add(payload.id)
+		}
+
+		onBeforeMount(() => {
+			field.currentValue.forEach((fileItem) => {
+				if (
+					![
+						undefined,
+						KottiFieldFileUploadRemote.Status.UPLOADED,
+						KottiFieldFileUploadRemote.Status.UPLOADED_WITH_ERROR,
+					].includes(fileItem.status)
+				)
+					throw new Error(
+						`KtFieldFileUpload: invalid status for file item ${fileItem.id}: ${fileItem.status}`,
+					)
+
+				preUploadedFiles.add(fileItem.id)
+			})
+		})
+
+		watch(
+			() => props.payload,
+			(newPayload, oldPayload) => {
+				Object.entries(newPayload).forEach(([id, { status }]) => {
+					const oldStatus = oldPayload[id]?.status
+
+					if (oldStatus && status === oldStatus) return
+
+					const fileItemStatus =
+						field.currentValue.find((fileItem) => fileItem.id === id)?.status ??
+						null
+
+					if (
+						fileItemStatus !== null &&
+						![
+							KottiFieldFileUploadRemote.Status.INVALID,
+							KottiFieldFileUploadRemote.Status.NOT_STARTED,
+							KottiFieldFileUploadRemote.Status.UPLOADED,
+							KottiFieldFileUploadRemote.Status.UPLOADED_WITH_ERROR,
+						].includes(fileItemStatus)
+					)
+						setStatus({ id, status })
+				})
+			},
+			{ deep: true },
```

Comment:
This will allow deep mutating from outside rather than replacing the payload object all together whenever a remote is updated

Did you test without this?

Reply:
Tested, works without this. It's removed now.